% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MetaSTAAR_merge_cond.R
\name{MetaSTAAR_merge_cond}
\alias{MetaSTAAR_merge_cond}
\title{The preliminary data manipulation step for \code{MetaSTAAR_cond}}
\usage{
MetaSTAAR_merge_cond(
  chr,
  start.loc,
  end.loc,
  study.names,
  sample.sizes,
  sumstat.dir,
  cov.dir,
  covcond.dir,
  rare_maf_cutoff = 0.01,
  cov_maf_cutoff,
  trait,
  region,
  segment.size = 5e+05,
  effect.cond = c("homogeneous", "heterogeneous"),
  check_qc_label = FALSE
)
}
\arguments{
\item{chr}{a numeric value indicating the chromosome of the genetic region of interest.}

\item{start.loc}{a numeric value indicating the starting location (position) of the
genetic region of interest.}

\item{end.loc}{a numeric value indicating the ending location (position) of the
genetic region of interest.}

\item{study.names}{a character vector containing the name of each participating
study in the meta-analysis.}

\item{sample.sizes}{a numeric vector with the length of \code{study.names}
indicating the sample size of each study.}

\item{sumstat.dir}{a character vector containing the directories of the study-specific summary statistics file folders.}

\item{cov.dir}{a character vector containing the directories of the study-specific sparse weighted
covariance file folders.}

\item{covcond.dir}{a character vector containing the directories of the study-specific covariance file folders
for conditional analysis.}

\item{rare_maf_cutoff}{the cutoff of maximum minor allele frequency in
defining rare variants (default = 0.01).}

\item{cov_maf_cutoff}{a numeric vector with the length of \code{study.names}
indicating the maximum minor allele frequency cutoffs under which the sparse weighted
covariance files between variants are stored.}

\item{trait}{a character value indicating the underlying trait of interest for
the meta-analysis.}

\item{region}{a character value indicating the underlying region of a given list
of variants adjusted for conditional analysis of the meta-analysis.}

\item{segment.size}{a numeric value indicating the length of each segment of which
the summary statistics and sparse weighted covariance files are stored.
Note that the input value should be aligned with the input values of
\code{\link{MetaSTAAR_worker_cov}} (default = 5e+05).}

\item{effect.cond}{a character value indicating the effects of variants to be adjusted for
in conditional analysis are "homogeneous" or "heterogeneous" (default = "homogeneous").}

\item{check_qc_label}{a logical value indicating whether variants need to be dropped according to \code{qc_label}
specified in \code{\link{MetaSTAAR_worker_sumstat}} and \code{\link{MetaSTAAR_worker_cov}}.
If \code{check_qc_label} is FALSE, it is assumed that no variant will be dropped (default = FALSE).}
}
\value{
a list with the following members:

\code{info}: the merged data frame of all variants in the genetic region
of interest whose combined minor allele frequency is below \code{rare_maf_cutoff}, including the
following information (listed in the same order as \code{U} and the rows/columns of \code{cov}):
chromosome (chr), position (pos), reference allele (ref), alternative allele (alt),
combined minor allele count (MAC), and combined minor allele frequency (MAF).

\code{U_cond}: the merged conditional score statistics vector of all variants in the genetic region
of interest whose combined minor allele frequency is below \code{rare_maf_cutoff},
adjusting for a given list of variants.

\code{cov_cond}: the merged conditional covariance matrix of all variants in the genetic region
of interest whose combined minor allele frequency is below \code{rare_maf_cutoff},
adjusting for a given list of variants.
}
\description{
The \code{MetaSTAAR_merge_cond} function takes in the summary statistics file and the sparse weighted
covariance file (the output from \code{\link{MetaSTAAR_worker_sumstat}} and \code{\link{MetaSTAAR_worker_cov}})
as well as the covariance file for conditional analysis (the output from \code{\link{MetaSTAAR_worker_cov_cond}})
from each participating study and performs the preliminary data manipulation step
by merging them into a single unified conditional summary statistics
file and a conditional covariance file, respectively, adjusting for a given list of variants.
}
\references{
Li, X., et al. (2023). Powerful, scalable and resource-efficient
meta-analysis of rare variant associations in large whole genome sequencing studies.
\emph{Nature Genetics}, \emph{55}(1), 154-164.
(\href{https://doi.org/10.1038/s41588-022-01225-6}{pub})
}
